### PR TITLE
[Add] Function calling support for Autobuild

### LINKED ISSUE #1510: [Feature Request]: how do we add tools to agents created by AgentBuilder ?
### Is your feature request related to a problem? Please describe.

I want to integrate my custom functions and langchain tools with agents created by AgentBuilder

### Describe the solution you'd like

It should be possible to use custom functions or langchain's tools by these agents, As other UserProxyAgent, GPTAssistantAgent and many have the option to. If this is a feature on which work needs to be started, I can also take it up 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by krishnashed**
## Why are these changes needed?

Introduction of Autobuild, didn't support function/tool calling. This PR is solution to add this feature/

## Related issue number

Closes #1510 

## Checks
Check and tests still need to be written and passed.

### CODE REVIEW

**Comment by krishnashed** [comment]
@microsoft-github-policy-service agree

**Comment by ekzhu** [comment]
You can fix code formatting running `pre-commit run --all-files` locally.

**Comment by jamesliu** [comment]
@krishnashed, I'd like to explore and validate any test scenarios you might have. Thanks.

**Comment by JieyuZ2** [comment]
hey @krishnashed very nice feature, could you check this feature and make it compatible? https://microsoft.github.io/autogen/docs/topics/code-execution/user-defined-functions/

**Comment by krishnashed** [comment]
Sure @JieyuZ2 I'll check it out and make sure it's compatible

**Comment by krishnashed** [comment]
@JieyuZ2 @jamesliu I have added support for OpenAI Assistant generated using AutoBuild to access function calling. Also added working ipynb files for same

File: OAI_CONFIG_LIST_sample [hunk 1]
```diff
@@ -21,4 +21,4 @@
         "api_type": "azure",
         "api_version": "2024-02-15-preview"
     }
-]
+]
```
Comment (sonichi) [inline OAI_CONFIG_LIST_sample:24]: Remove this change.
```suggestion
]

```

**Review by krishnashed — COMMENTED** [review]
Closing it as its implemented, and pre-build checks are now passing

**Comment by krishnashed** [comment]
@jamesliu I have added test cases for Autobuild function calling, can you please review?

File: autogen/agentchat/contrib/agent_builder.py [hunk 2]
```diff
@@ -642,13 +686,58 @@ def _build_agents(
 
         Returns:
             agent_list: a list of agents.
+            list_of_functions: List of functions to be registered with agents.
```
Comment (jamesliu) [inline autogen/agentchat/contrib/agent_builder.py:689]: list_of_functions in Args, not in Returns

**Pull request closed without merging**